File: src/domain/market/player_valuation.ts
import type { Player } from "@/domain/player/player";

// Market valuation of a player. Decoupled from Player identity so the pricing
// source (mock today, Sportmonks tomorrow) can evolve independently of the
// player reference data.

export type ValuationSource = "mock" | "sportmonks" | "synthetic" | "engine";

export interface PlayerValuation {
  player_id: number;
  base_value: number; // €M, reference value (tournament-open anchor)
  current_price: number; // €M, latest tradable price
  change_since_inception: number; // %, current vs base — the canonical "% change" (screeners / movers)
  change_avg_per_match: number; // %, mean net change per fixture priced
  change_last_match: number; // %, net change over the latest fixture — moves live during play
  performance_rating: number; // 0-10, latest match rating
  as_of: string; // ISO timestamp of the snapshot
  source: ValuationSource;
}

// Convenience join used by surfaces that always need both — screener rows,
// movers lists, watchlist tiles. UI layer reads it like one object.
export interface PlayerWithValuation extends Player {
  valuation: PlayerValuation;
}

File: backend/src/domain/valuation/player_valuation.py
"""PlayerValuation — Value Object representing a price snapshot for a player.

DDD role: Value Object (no identity beyond player_id + as_of). Produced by the
valuation engine (M5) or by a stub provider (M3).
"""

from dataclasses import dataclass
from datetime import datetime
from enum import StrEnum


class ValuationSource(StrEnum):
    SYNTHETIC = "synthetic"  # M3 placeholder: deterministic pseudo-random
    ENGINE = "engine"  # M5+: output of XGBasedStrategyV0
    REHEARSAL = "rehearsal"  # synthetic-rating UI/UX rehearsal — NEVER real data
    SETTLEMENT = "settlement"  # full-time tournament RESULT event (win / elimination)
    QUALIFICATION = "qualification"  # group qualification (team reaches the knockout bracket)
    SUSPENSION = "suspension"  # banned for the next match (red / 2-yellow accumulation)
    LINEUP_DROP = "lineup_drop"  # dropped from the XI (started last match, benched this one)


@dataclass(frozen=True, slots=True)
class PlayerValuation:
    """All ``change_*`` values are percentages (e.g. ``5.0`` means +5%)."""

    player_id: int
    base_value: float
    current_price: float
    # Cumulative return vs the tournament-open price. The canonical "% change"
    # for screeners / top-movers. ((current_price / base_value) - 1) * 100.
    # Always a number: an un-played player legitimately sits at 0% (price = base).
    change_since_inception: float
    # Mean, over every fixture the player has been priced in, of that fixture's
    # net price change. None when the player has not been priced in any fixture
    # (no matches yet → "n/a", NOT 0%, which would read as "a flat match").
    change_avg_per_match: float | None
    # Net price change over the most recent fixture the player was priced in —
    # i.e. the latest tick's "change since that fixture's kickoff". Moves live
    # while a match is in progress, then holds until the next match's first tick.
    # None when the player has no fixture ticks yet (same n/a vs 0% reasoning).
    change_last_match: float | None
    performance_rating: float
    as_of: datetime
    source: ValuationSource
